spotbugs/src/main/java/edu/umd/cs/findbugs/util/MultiThreadedCodeIdentifierUtils.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.bcel.classfile.Utility;
+import org.apache.bcel.generic.ConstantPoolGen;
+import org.apache.bcel.generic.INVOKEVIRTUAL;
+import org.apache.bcel.generic.Instruction;
+import org.apache.bcel.generic.InstructionHandle;
+import org.apache.bcel.generic.MONITORENTER;
+import org.apache.bcel.generic.MethodGen;
+
+/**
+ * Utility class with methods to identify multi threaded code
+ */
+public class MultiThreadedCodeIdentifierUtils {
+
+    private static final String RUNNABLE_SIGNATURE = "java.lang.Runnable";
+    private static final String ATOMIC_PACKAGE = "java.util.concurrent.atomic";
+
+    public static boolean isPartOfMultiThreadedCode(ClassContext classContext) {
+        JavaClass classToCheck = classContext.getJavaClass();
+        if (implementsRunnable(classToCheck) ||
+                Stream.of(classToCheck.getFields()).anyMatch(MultiThreadedCodeIdentifierUtils::isFieldMultiThreaded)) {
+            return true;
+        }
+
+        return Stream.of(classToCheck.getMethods()).anyMatch(method -> isMethodMultiThreaded(method, classContext));
+    }
+
+    public static boolean implementsRunnable(JavaClass javaClass) {
+        try {
+            return Stream.of(javaClass.getAllInterfaces())
+                    .anyMatch(ifc -> ifc.getClassName().equals(RUNNABLE_SIGNATURE));
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    public static boolean isMethodMultiThreaded(Method method, ClassContext classContext) {
+        if (method.isSynchronized()) {
+            return true;
+        }
+
+        LocalVariableTable lvt = method.getLocalVariableTable();
+        if (lvt != null && Stream.of(lvt.getLocalVariableTable())
+                .anyMatch(lv -> signatureIsFromAtomicPackage(lv.getSignature()))) {
+            return true;
+        }
+
+        Optional<MethodGen> maybeMethodGen = Optional.ofNullable(classContext.getMethodGen(method));
+        if (!maybeMethodGen.isPresent()) {
+            return false;
+        }
+
+        MethodGen methodGen = maybeMethodGen.get();
+        return hasMultiThreadedInstruction(methodGen);
+    }
+
+    public static boolean hasMultiThreadedInstruction(MethodGen methodGen) {
+        ConstantPoolGen cpg = methodGen.getConstantPool();
+        InstructionHandle handle = methodGen.getInstructionList().getStart();
+        while (handle != null) {
+            Instruction ins = handle.getInstruction();
+            if (ins instanceof MONITORENTER) {
+                return true;
+            } else if (ins instanceof INVOKEVIRTUAL) {
+                String methodName = ((INVOKEVIRTUAL) ins).getMethodName(cpg);
+                if (isConcurrentLockInterfaceCall(methodName)) {
+                    return true;
+                }
+            }
+            handle = handle.getNext();
+        }
+
+        return false;
+    }
+
+    public static boolean isConcurrentLockInterfaceCall(String methodName) {
```

Comment:
As far as I can see, this function is only called from inside this class. Wouldn't it be better to restrict the visibility?